enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,213 @@
+2. microshift-storage-topolvm
+
+### Goals
+* Allow networking, storage and router core components to be optionally
+  installed.
+
+* As we refactor and re-use the kustomize package used for loading user workloads,
+  add support for loading yamls from multiple directories, i.e.:
+  * `/usr/lib/microshift/manifests.d/user-app-1/`
+  * `/usr/lib/microshift/manifests.d/user-app-2/`
+  * `/etc/microshift/manifests.d/user-app-3/`
+  * `...`
+
+* Split the RPM packages to resemble those optional components.
+
+* Preserve previous behavior in terms of RPM package installation.
+
+### Non-Goals
+
+* Making DNS component optional.
+* Making the certificate management component optional.
+
+## Proposal
+
+To allow modularity of the described core components, the following changes
+would be introduced:
+
+Update the service manager to allow core components as kustomization yamls,
+in the same way we already support for user application manifests. i.e.:
+
+`/usr/lib/microshift/components.d/networking-ovn/kustomize.yaml`
+`/usr/lib/microshift/components.d/storage-topolvm/kustomize.yaml`
+`/usr/lib/microshift/components.d/openshift-router/kustomize.yaml`
+
+  MicroShift component manager would scan the `/usr/lib/microshift/components.d/`
+  directory and extract metadata from the yamls to analyze dependencies between
+  core components at boot. Then start the right componments in the right order.
+
+* The kustomize component loader would be refactored so it can be used for this
+  purpose. We should be able to pre-process kustomize yamls to insert template
+  variables as we do today for core assets.
+
+The following changes would be introduced to packaging:
+
+* Convert `microshift` into a virtual package that depends on `microshift-core`,
+  `microshift-networking-ovn`, `microshift-storage-topolvm`, `microshift-openshift-router`
+  to preserve the current behavior.
+
+* Convert the `microshift-networking` package would be made a virtual package which
+  depends on `microshift-networking-ovn` to preserver current behavior.
+
+* The `microshift-networking-ovn` package would include the necessary
+  `/usr/lib/microshift/components.d/networking-ovn/` yamls, including crds.
+
+* The `microshift-storage-topolvm` package would include the necessary
+  `/usr/lib/microshift/components.d/storage-topolvm` yamls, including crds.
+
+* The `microshift-openshift-router` package would include the necessary
+  `/usr/lib/microshift/components.d/openshift-router` yamls, including crds.
+
+The rebase script should be updated to extract those core components
+as kustomize yamls instead of embedded assets.
+
+### Workflow Description
+
+#### Deploying
+
+The workflow for deployment is already described in the user histories.
+
+#### Upgrading
+
+Upgrading optional components would behave in the same way as upgrading
+components where the assets are embedded into the MicroShift binary.
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
```

Comment:
Good question.

I will investigate that, I guess we will need to keep parsing for those files in MicroShift, or generalize that parsing more, so then it can be injected via the templates.

Reply:
Do you think my proposal bellow would make sense?

Reply:
@zshi-redhat please have a look on what I proposed here, let me know if you think it would make sense.